presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java

```diff
@@ -0,0 +1,173 @@
+                .map(Map.Entry::getValue)
+                .flatMap(e -> DependencyExtractor.extractAll(e).stream())
+                .collect(toSet());
+
+        Assignments.Builder childAssignments = Assignments.builder();
+        for (Map.Entry<Symbol, Expression> assignment : child.getAssignments().entrySet()) {
+            if (!candidates.contains(assignment.getKey())) {
+                childAssignments.put(assignment.getKey(), assignment.getValue());
+            }
+        }
+        for (Symbol input : inputs) {
+            childAssignments.put(input, input.toSymbolReference());
+        }
+
+        return Optional.of(
+                new ProjectNode(
+                        parent.getId(),
+                        new ProjectNode(
+                                child.getId(),
+                                child.getSource(),
+                                childAssignments.build()),
+                        Assignments.copyOf(parentAssignments)));
+    }
+
+    private Sets.SetView<Symbol> extractCandidates(ProjectNode parent, ProjectNode child)
+    {
+        // candidates for inlining are
+        //   1. references to simple constants
+        //   2. references to complex expressions that
+        //      a. are not inputs to try() expressions
+        //      b. appear only once across all expressions
+        //      c. are not identity projections
+
+        Map<Symbol, Long> dependencies = parent.getAssignments()
+                .getExpressions().stream()
+                .flatMap(e -> DependencyExtractor.extractAll(e).stream())
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+
+        // find references to simple constants
+        Set<Symbol> constants = dependencies.keySet().stream()
+                .filter(input -> child.getAssignments().get(input) instanceof Literal)
+                .collect(toSet());
+
+        // exclude any complex inputs to TRY expressions. Inlining them would potentially
+        // change the semantics of those expressions
+        Set<Symbol> tryArguments = parent.getAssignments()
+                .getExpressions().stream()
+                .flatMap(e -> extractTryArguments(e).stream())
+                .collect(toSet());
+
+        Set<Symbol> singletons = dependencies.entrySet().stream()
+                .filter(e -> e.getValue() == 1) // reference appears just once across all expressions in parent project node
+                .filter(e -> !tryArguments.contains(e.getKey())) // they are not inputs to TRY. Otherwise, inlining might change semantics
+                .filter(e -> !isIdentity(e.getKey(), child.getAssignments().get(e.getKey()))) // skip identities, otherwise, this rule will keep firing forever
+                .map(Map.Entry::getKey)
+                .collect(toSet());
+
+        return Sets.union(singletons, constants);
+    }
+
+    private boolean isIdentity(Symbol output, Expression input)
```

Comment:
IMO this should belong to `Assignments` as it full name should be `isIdentityAssignment`, what do you think?